### PR TITLE
Persist docker port mapping in guestinfo

### LINKED ISSUE #2452: Map port for existing containers after VCH restart
For containers used port mapping, need to setup ip table for those containers during container cache population.

### PR DESCRIPTION
**Pull request by emlin**
Fixes #2452 

Persist docker port mapping information to make it survive VCH restart.
The information is in guestinfo right now, to avoid performance issue caused by current kv store.

### CODE REVIEW

File: tests/test-cases/Group10-VCH-Restart/Group10-VCH-Restart.robot [hunk 1]
```diff
@@ -28,6 +28,18 @@ Created Network Persists And Containers Are Discovered With Correct IPs
...
     ${id1}  ${ip1}=  Launch Container  vch-restart-test1  bridge
     ${id2}  ${ip2}=  Launch Container  vch-restart-test2  bridge
+
+    ${rc}  ${output}=  Run And Return Rc And Output  docker ${params} create -it -p 10000:80 -p 10001:80 --name webserver nginx
+    Should Be Equal As Integers  ${rc}  0
+    Should Not Contain  ${output}  Error
+    ${rc}  ${output}=  Run And Return Rc And Output  docker ${params} start webserver
+    Should Be Equal As Integers  ${rc}  0
+    Should Not Contain  ${output}  Error
+    ${rc}  ${output}=  Run And Return Rc And Output  wget ${vch-ip}:10000 --retry-connrefused --timeout=10 --tries=10
```
Comment (anchal-agrawal) [inline tests/test-cases/Group10-VCH-Restart/Group10-VCH-Restart.robot]: These can be refactored using a recent change: https://github.com/vmware/vic/blob/master/tests/test-cases/Group1-Docker-Commands/1-25-Docker-Port-Map.robot#L8

File: lib/portlayer/network/endpoint.go [hunk 2]
```diff
@@ -31,14 +31,15 @@ type alias struct {
...
-	static    bool
-	ports     map[Port]interface{} // exposed ports
-	aliases   map[string][]alias
+	container   *Container
+	scope       *Scope
+	ip          net.IP
+	gateway     net.IP
+	subnet      net.IPNet
+	static      bool
+	ports       map[Port]interface{} // exposed ports
```
Comment (hmahmood) [inline lib/portlayer/network/endpoint.go]: Is there a need for this with `portmapping` below?

Comment (emlin) [inline lib/portlayer/network/endpoint.go]: This map is still used to check exposed ports, to see if port is already consumed by other mapping.
The new attribute below is just for persistence. I don't think portlayer should care too much about the conversion from nat.Portbinding to string.

File: lib/apiservers/engine/backends/container_proxy_test.go [hunk 3]
```diff
@@ -47,3 +48,31 @@ func TestProcessVolumeParams(t *testing.T) {
 	invalidFields, _ := processVolumeParam(invalidVolume)
 	assert.Equal(t, volumeFields{}, invalidFields)
 }
+
+func TestPort(t *testing.T) {
+
+	portMap, bindingMap, err := nat.ParsePortSpecs([]string{"1236:1235/tcp", "1237:1235/tcp", "2345/udp", "80", "127.0.0.1::8080", "127.0.0.1:5279:8080"})
+	if err != nil {
+		t.Errorf("Failed to part ports: %s", err.Error())
```
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/container_proxy_test.go]: Should this be `parse ports`?

File: tests/resources/Util.robot [hunk 4]
```diff
@@ -436,3 +436,8 @@ Put Host Into Maintenance Mode
 Remove Host From Maintenance Mode
     ${rc}  ${output}=  Run And Return Rc And Output  govc host.maintenance.exit -host.ip=%{TEST_URL}
     Should Contain  ${output}  exiting maintenance mode... OK
+
+Hit Nginx Endpoint
+    [Arguments]  ${vch-ip}  ${port}
+    ${rc}  ${output}=  Run And Return Rc And Output  wget ${vch-ip}:${port}
+Should Be Equal As Integers ${rc} 0
```
Comment (anchal-agrawal) [inline tests/resources/Util.robot]: Needs indentation.

File: lib/apiservers/engine/backends/backends.go [hunk 5]
```diff
@@ -197,18 +202,52 @@ func InsecureRegistries() []string {
...
+		return errors.Errorf("Failed to set port mapping: %s", strings.Join(errs, "\n"))
+	}
+	return nil
+}
+
+func setPortMapping(backend *Container, container *container.VicContainer) error {
+	log.Debugf("Set port mapping for container %q", container.Name)
+
+	client := backend.containerProxy.Client()
+	endpointsOK, err := client.Scopes.GetContainerEndpoints(scopes.NewGetContainerEndpointsParamsWithContext(ctx).WithHandleOrID(container.ContainerID))
```
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/backends.go]: Consider breaking this up into separate lines for readability.

File: lib/apiservers/engine/backends/container_proxy_test.go [hunk 6]
```diff
@@ -47,3 +48,31 @@ func TestProcessVolumeParams(t *testing.T) {
 	invalidFields, _ := processVolumeParam(invalidVolume)
 	assert.Equal(t, volumeFields{}, invalidFields)
 }
+
+func TestPort(t *testing.T) {
+
+	portMap, bindingMap, err := nat.ParsePortSpecs([]string{"1236:1235/tcp", "1237:1235/tcp", "2345/udp", "80", "127.0.0.1::8080", "127.0.0.1:5279:8080"})
```
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/container_proxy_test.go]: Minor comment - separating this into multiple lines would improve readability.

File: lib/portlayer/network/endpoint.go (hunk 2 above)
Comment (emlin) [inline lib/portlayer/network/endpoint.go]: There is no 1X1 mapping for the exposed ports and portbinding string, for example, the string looks like "80:80/tcp", "8080:80/tcp", but the exposed ports will be "80/tcp" only.

portlayer will check "80/tcp" only, but persistence need to keep "80:80/tcp", "8080:80/tcp". 
Ideally portlayer should not care about whole portmapping, if we merge them two, we'll need to have portlayer convert the nat.Map back to string, and in the future, if we move the portmapping to kv store, we'll need to separate them again.

Comment (emlin) [inline lib/portlayer/network/endpoint.go]: OK, get rid of this, cause it doesn't work for not started container.

File: lib/apiservers/engine/backends/backends.go [hunk 7]
```diff
@@ -197,18 +201,56 @@ func InsecureRegistries() []string {
...
+	}
+	if *info.ContainerConfig.State != "Running" || len(container.HostConfig.PortBindings) == 0 {
+		log.Infof("Container info state: %s", *info.ContainerConfig.State)
+		log.Infof("container portbinding: %+v", container.HostConfig.PortBindings)
+		return nil
+	}
+
+	log.Debugf("Set port mapping for container %q, portmapping %+v", container.Name, container.HostConfig.PortBindings)
+	client := backend.containerProxy.Client()
+	endpointsOK, err := client.Scopes.GetContainerEndpoints(
```
Comment (hmahmood) [inline lib/apiservers/engine/backends/backends.go]: Do we need this call, given that the full port mapping info is in HostConfig?

Comment (emlin) [inline lib/apiservers/engine/backends/backends.go]: this call is to get back endpoint information, including the containerIP, which is not persisted in portmapping.
Need this information if container is running

Comment (hmahmood) [inline lib/apiservers/engine/backends/backends.go]: Why do you need IP information at this stage? This should only be collected when `inspect` is called.

Comment (emlin) [inline lib/apiservers/engine/backends/backends.go]: But the method backend.mapPorts require that information to map port.

**Comment by hmahmood** [comment]
lgtm

**Comment by anchal-agrawal** [comment]
:shipit:

**Comment by emlin** [comment]
⏺

**Pull request merged**